src/metabase/util/password.clj

```diff
@@ -47,17 +50,41 @@
                  (when (>= (occurances char-type) min-count)
                    (recur more)))))))
 
-(def active-password-complexity
+(defn active-password-complexity
   "The currently configured description of the password complexity rules being enforced"
+  []
   (merge (complexity->char-type->min (config/config-kw :mb-password-complexity))
          ;; Setting MB_PASSWORD_LENGTH overrides the default :total for a given password complexity class
          (when-let [min-len (config/config-int :mb-password-length)]
            {:total min-len})))
 
-(def ^{:arglists '([password])} is-complex?
+(defn- is-complex?
   "Check if a given password meets complexity standards for the application."
-  (partial password-has-char-counts? active-password-complexity))
+  [password]
+  (password-has-char-counts? (active-password-complexity) password))
+
+(def ^File common-passwords-file
+  "A set of ~12k common passwords to reject, that otherwise meet Metabase's default complexity requirements.
+   Sourced from Dropbox's zxcvbn repo: https://github.com/dropbox/zxcvbn/blob/master/data/passwords.txt"
+  (io/file (io/resource "common_passwords.txt")))
+
+(defn- is-uncommon?
+  "Check if a given password is not present in the common passwords set. Case-insensitive search since
+   the list only contains lower-case passwords."
+  [password]
+  (with-open [reader (java.io.BufferedReader. (java.io.FileReader. common-passwords-file))]
+    (not-any?
+      (partial = (str/lower-case password))
+      (iterator-seq (.. reader lines iterator)))))
 
+(def ^{:arglists '([password])} is-valid?
+  "Check that a password both meets complexity standards, and is not present in the common passwords list.
+   Common password list is ignored if minimum password complexity is set to :weak"
+  (every-pred
+    is-complex?
+    #(or
+       (= :weak (config/config-kw :mb-password-complexity))
+       (is-uncommon? %))))
```

Comment:
A couple pointers here -- using `def` and `every-pred` is a lot weirder to look at then just doing a normal `defn` like this:

```clj
(defn is-valid? [password]
  (and (is-complex? password)
       (or (= (config/config-kw :mb-password-complexity) :week)
           (is-uncommon? password))))
```

`comp`, `partial`, `every-pred`, and the like have uses but be mindful about being too clever with them at the expensive or readability.

Another pointer:

```clj
(= :weak (config/config-kw :mb-password-complexity))
``` 

is a [Yoda Condition](https://knowthecode.io/yoda-conditions-yoda-not-yoda). It would be like writing the equivalent JavaScript

```js
'week' === Config.keyword('mb-password-complexity')
```

It's not _wrong_, but it's a little weird. An easy way to think about this with Lisp languages is to mentally move the prefix operator to the equivalent infix position(s). e.g.

```
(/ x y) => x / y
(> a b c) => a > b > c
```